### PR TITLE
Aero Sanity update

### PR DESCRIPTION
**Pull request by mkerensky**
+ Aero Sanity 2.0!
+ Allow targeting for waypoint launch bearings-only capital missile bay modes
+ Remove spurious Capital Missile critical roll for thunderbolt type missiles

This is an update and rework of aero sanity. Fixes a number of reporting bugs and adds full support for Advanced Point Defense, Glancing/Direct blows, S2S nuclear weapons and Bearings-Only Capital Missiles.
This supercedes PR#1190.

### CODE REVIEW

File: megamek/src/megamek/common/weapons/AR10Handler.java [hunk 1]
```diff
@@ -38,6 +46,229 @@
...
+        if (!bMissed && (entityTarget != null)) {
+            handleEntityDamage(entityTarget, vPhaseReport, bldg, hits,
+                    nCluster, bldgAbsorbs);
+            server.creditKill(entityTarget, ae);
+        } else if (!bMissed) { // Hex is targeted, need to report a hit
+            r = new Report(3390);
+            r.subject = subjectId;
+            vPhaseReport.addElement(r);
+        }
+        }
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/AR10Handler.java]: whitespace is off

**Comment by HammerGS** [comment]
This pull request **introduces 12 alerts** when merging 9e677be54b404403d38f2ec4cc86ddb0256c36de into 028a56398e441a1657488d076a43ab758e3aadca - [view on LGTM.com](https://lgtm.com/projects/g/MegaMek/megamek/rev/pr-d34232d4cea31fb1d8709b31354dd25d052fae4f)

**new alerts:**

* 11 for Dereferenced variable may be null
* 1 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: megamek/src/megamek/common/weapons/BayWeaponHandler.java [hunk 2]
```diff
@@ -498,8 +540,13 @@ public boolean handleAeroSanity(IGame.Phase phase, Vector<Report> vPhaseReport)
                 if(bayWType instanceof Weapon) {
                     replaceReport = vPhaseReport.size();
                     WeaponAttackAction bayWaa = new WeaponAttackAction(waa.getEntityId(), waa.getTargetType(), waa.getTargetId(), wId);
-                    AttackHandler bayWHandler = ((Weapon)bayWType).getCorrectHandler(toHit, bayWaa, game, server);
+                    AttackHandler bayWHandler = ((Weapon)bayWType).getCorrectHandler(autoHit, bayWaa, game, server);
                     bayWHandler.setAnnouncedEntityFiring(false);
+                    // This should always be true. Maybe there's a better way to write this?
+                    if (bayWHandler instanceof WeaponHandler) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/BayWeaponHandler.java]: as in `bayWHandler` is always a `WeaponHandler`?

File: megamek/src/megamek/common/weapons/CLIATMHandler.java [hunk 3]
```diff
@@ -270,14 +270,21 @@ protected int calcMissileHits(Vector<Report> vPhaseReport) {
         // add AMS mods
         int amsMod = getAMSHitsMod(vPhaseReport);
         nMissilesModifier += amsMod;
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/CLIATMHandler.java]: `entityTarget` may be null.

File: megamek/src/megamek/common/weapons/LRMAntiTSMHandler.java [hunk 4]
```diff
@@ -76,8 +79,21 @@ protected int calcHits(Vector<Report> vPhaseReport) {
         
         // AMS mod
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/LRMAntiTSMHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/LRMHandler.java [hunk 5]
```diff
@@ -264,19 +265,30 @@ protected int calcHits(Vector<Report> vPhaseReport) {
 
         // add AMS mods
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)
+                && entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/LRMHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/LRMSwarmHandler.java [hunk 6]
```diff
@@ -448,13 +448,23 @@ protected int calcHits(Vector<Report> vPhaseReport) {
         int nMissilesModifier = getClusterModifiers(false);
 
         // add AMS mods
-        nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        int amsMod = getAMSHitsMod(vPhaseReport);
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/LRMSwarmHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/MissileWeaponHandler.java [hunk 7]
```diff
@@ -245,9 +243,18 @@ protected int calcHits(Vector<Report> vPhaseReport) {
 
         // add AMS mods
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)
+                && entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/MissileWeaponHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/NarcHandler.java [hunk 8]
```diff
@@ -59,7 +61,20 @@ public NarcHandler(ToHitData t, WeaponAttackAction w, IGame g, Server s) {
     protected int calcHits(Vector<Report> vPhaseReport) {
         bSalvo = true;
         getAMSHitsMod(vPhaseReport);
-        calcCounterAV();
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            // Or bay AMS if Aero Sanity is on
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/NarcHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/SRMAntiTSMHandler.java [hunk 9]
```diff
@@ -69,8 +72,21 @@ protected int calcHits(Vector<Report> vPhaseReport) {
       
         // Add ams mod
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/SRMAntiTSMHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/SRMInfernoHandler.java [hunk 10]
```diff
@@ -230,9 +230,21 @@ protected int calcHits(Vector<Report> vPhaseReport) {
 
         // add AMS mods
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/SRMInfernoHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/StreakHandler.java [hunk 11]
```diff
@@ -103,6 +105,15 @@ protected int calcHits(Vector<Report> vPhaseReport) {
 
         int missilesHit;
         int amsMod = getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/StreakHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/ThunderBoltWeaponHandler.java [hunk 12]
```diff
@@ -125,7 +129,20 @@ protected boolean usesClusterTable() {
      */
     @Override
     protected int calcHits(Vector<Report> vPhaseReport) {
+        // Activate single AMS
         getAMSHitsMod(vPhaseReport);
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)) {
+            // Or bay AMS if Aero Sanity is on
+            Entity entityTarget = (target.getTargetType() == Targetable.TYPE_ENTITY) ? (Entity) target
+                    : null;
+            if (entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/ThunderBoltWeaponHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/ATMHandler.java [hunk 13]
```diff
@@ -289,9 +296,18 @@ protected int calcStandardAndExtendedAmmoHits(Vector<Report> vPhaseReport) {
 
         // add AMS mods
         nMissilesModifier += getAMSHitsMod(vPhaseReport);
+        
+        if (game.getOptions().booleanOption(OptionsConstants.ADVAERORULES_AERO_SANITY)
+                && entityTarget.isLargeCraft()) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/ATMHandler.java]: `entityTarget` may be null

File: megamek/src/megamek/common/weapons/WeaponHandler.java [hunk 14]
```diff
@@ -367,6 +378,24 @@ protected int getCounterAV() {
...
+     */ 
+    protected WeaponHandler getParentBayHandler() {
+        return parentBayHandler;
+    }
+    
+    /**
+     * Sets the parent handler for each sub-weapon handler called when looping through bay weapons
+     * Used with Aero Sanity to pass counterAV through to the individual missile handler from the bay handler
+     * 
+     * @param ah - The <code>AttackHandler</code> for the BayWeapon this individual weapon belongs to
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/WeaponHandler.java]: `bh` is the parameter

File: megamek/src/megamek/common/weapons/CapitalMissileHandler.java [hunk 15]
```diff
@@ -47,6 +54,252 @@ public CapitalMissileHandler(ToHitData t, WeaponAttackAction w, IGame g,
...
+            } else {
+                r.messageId = 3120;
+                r.add(target.getDisplayName(), true);
+            }
+            vPhaseReport.addElement(r);
+                
+        
+        
+        // are we a glancing hit?  Check for this here, report it later
+        if (game.getOptions().booleanOption(OptionsConstants.ADVCOMBAT_TACOPS_GLANCING_BLOWS)) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/CapitalMissileHandler.java]: whitespace is off at this point

File: megamek/src/megamek/common/weapons/WeaponHandler.java [hunk 16]
```diff
@@ -557,6 +599,83 @@ protected int calcnClusterAero(Entity entityTarget) {
...
                         vPhaseReport.add(r);
                     } else {
+                        //If point defenses engage Large, single missiles
+                        if (pdBayEngagedMissile || amsBayEngagedMissile) {
+                            // remove the last reports because they showed the
+                            // number of shots that hit
+                            while (vPhaseReport.size() > reportSize) {
+                                vPhaseReport.remove(vPhaseReport.size() - 1);
+                            }
+                            int AMSHits = 0;
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/WeaponHandler.java]: prefer this declaration to be closer to the `for` loop

File: megamek/src/megamek/common/weapons/BayWeaponHandler.java (hunk 2 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/BayWeaponHandler.java]: Yes

Comment (sixlettervariables) [reply megamek/src/megamek/common/weapons/BayWeaponHandler.java]: Then I'd add in an else clause which raises a warning if it isn't.

File: megamek/src/megamek/common/weapons/BayWeaponHandler.java [hunk 17]
```diff
@@ -542,10 +543,14 @@ public boolean handleAeroSanity(IGame.Phase phase, Vector<Report> vPhaseReport)
...
                     AttackHandler bayWHandler = ((Weapon)bayWType).getCorrectHandler(autoHit, bayWaa, game, server);
                     bayWHandler.setAnnouncedEntityFiring(false);
-                    // This should always be true. Maybe there's a better way to write this?
+                    // This should always be true
                     if (bayWHandler instanceof WeaponHandler) {
                         WeaponHandler wHandler = (WeaponHandler) bayWHandler;
                         wHandler.setParentBayHandler(this);
+                    } else {
+                        logDebug(METHOD_NAME,
+                                "bayWHandler is not a weapon handler! How did you manage that?");
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/BayWeaponHandler.java]: could you include the entity ID or any other applicable identifier to help locate this problem if it occurs

File: megamek/src/megamek/common/weapons/CapitalMissileBayHandler.java [hunk 18]
```diff
@@ -754,6 +755,10 @@ public boolean handleAeroSanity(IGame.Phase phase, Vector<Report> vPhaseReport)
                     if (bayWHandler instanceof WeaponHandler) {
                         WeaponHandler wHandler = (WeaponHandler) bayWHandler;
                         wHandler.setParentBayHandler(this);
+                    } else {
+                        logDebug(METHOD_NAME,
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/CapitalMissileBayHandler.java]: same as above

File: megamek/src/megamek/common/weapons/AR10Handler.java (hunk 1 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/AR10Handler.java]: Fixed

File: megamek/src/megamek/common/weapons/BayWeaponHandler.java (hunk 2 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/BayWeaponHandler.java]: Fixed

File: megamek/src/megamek/common/weapons/CLIATMHandler.java (hunk 3 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/CLIATMHandler.java]: Fixed

File: megamek/src/megamek/common/weapons/LRMAntiTSMHandler.java (hunk 4 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/LRMAntiTSMHandler.java]: fixed

File: megamek/src/megamek/common/weapons/LRMHandler.java (hunk 5 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/LRMHandler.java]: fixed

**Comment by HammerGS** [comment]
This pull request **introduces 12 alerts** when merging 574d4dd441f24959f3361e4f702ff03544d9454b into 5dc7570b5116f8657cb7e875dbf132f3592087ae - [view on LGTM.com](https://lgtm.com/projects/g/MegaMek/megamek/rev/pr-67ca05012f605d112a1c869482a539a2ceb93f1a)

**new alerts:**

* 11 for Dereferenced variable may be null
* 1 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: megamek/src/megamek/common/weapons/ThunderBoltWeaponHandler.java (hunk 12 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/ThunderBoltWeaponHandler.java]: fixed

File: megamek/src/megamek/common/weapons/ATMHandler.java (hunk 13 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/ATMHandler.java]: fixed

File: megamek/src/megamek/common/weapons/LRMSwarmHandler.java (hunk 6 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/LRMSwarmHandler.java]: fixed

File: megamek/src/megamek/common/weapons/MissileWeaponHandler.java (hunk 7 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/MissileWeaponHandler.java]: fixed

File: megamek/src/megamek/common/weapons/NarcHandler.java (hunk 8 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/NarcHandler.java]: fixed

File: megamek/src/megamek/common/weapons/SRMAntiTSMHandler.java (hunk 9 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/SRMAntiTSMHandler.java]: fixed

File: megamek/src/megamek/common/weapons/SRMInfernoHandler.java (hunk 10 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/SRMInfernoHandler.java]: fixed

File: megamek/src/megamek/common/weapons/StreakHandler.java (hunk 11 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/StreakHandler.java]: fixed

File: megamek/src/megamek/common/weapons/WeaponHandler.java (hunk 14 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/WeaponHandler.java]: oops.  fixed.

File: megamek/src/megamek/common/weapons/CapitalMissileHandler.java (hunk 15 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/CapitalMissileHandler.java]: fixed

File: megamek/src/megamek/common/weapons/WeaponHandler.java (hunk 16 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/WeaponHandler.java]: Fixed I think

File: megamek/src/megamek/common/weapons/BayWeaponHandler.java (hunk 17 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/BayWeaponHandler.java]: Fixed

File: megamek/src/megamek/common/weapons/CapitalMissileBayHandler.java (hunk 18 above)
Comment (mkerensky) [reply megamek/src/megamek/common/weapons/CapitalMissileBayHandler.java]: Fixed

**Pull request merged**